File: 2020/day_1.py
import os


def part1(numbers: list, input_sum: int) -> int:
    """Return product of two numbers in <numbers> that sum to <input_sum>"""
    numbers_set = set(numbers)
    for n in numbers:
        if input_sum - n in numbers_set:
            return n * (input_sum - n)


def part2(numbers: list, input_sum: int) -> int:
    """Return product of three numbers in <numbers> that sum to <input_sum>"""
    numbers_set = set(numbers)
    for n1 in numbers:
        for n2 in numbers:
            if n1 != n2 and input_sum - n1 - n2 in numbers_set:
                return n1 * n2 * (input_sum - n1 - n2)


def read_file(file_name: str) -> list:
    return [int(line) for line in
            open(os.path.join(os.path.dirname(__file__), file_name), "r")]


def test():
    numbers = read_file("inputs/input1_test.txt")
    input_sum = 2020
    assert part1(numbers, input_sum) == 514579
    assert part2(numbers, input_sum) == 241861950


def main():
    test()
    input_sum = 2020
    numbers = read_file("inputs/input1.txt")
    print(part1(numbers, input_sum))
    print(part2(numbers, input_sum))


if __name__ == "__main__":
    main()

File: 2021/day_1.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cassert>

/* Read and parse input file <filename> */
std::vector<int> read_file(std::string filename) {
    std::ifstream fin; 
    fin.open(filename);
    
    int input;
    std::vector<int> inputs;
    while (fin >> input) {
        inputs.push_back(input);
    }

    return inputs;
}

/* Count number of increases of consecutive measurements */
int num_of_increases(const std::vector<int> input) {
    int increases = 0;
    for (decltype(input.size()) i = 1; i < input.size(); ++i) {
        increases += input[i] > input[i - 1];
    }
    return increases;
}

/* Count number of increases of consecutive window sums */
int num_of_increases_window(const std::vector<int> input) {
    int increases = 0;

    // We need at least 3 measurements to calculate the first window
    if (input.size() >= 3) {
        int prev_window = input[0] + input[1] + input[2];
        
        // Calculate the next window sum
        for (decltype(input.size()) i = 1; i + 2 < input.size(); ++i) {
            int curr_window = input[i] + input[i + 1] + input[i + 2];  
            increases += curr_window > prev_window;

            // Save current window sum for next iteration
            prev_window = curr_window;
        }
    }

    return increases;
}

/* Test small cases */
void test() {
    std::vector<int> test_inputs = read_file("inputs/input1_test.txt");
    assert(num_of_increases(test_inputs) == 7);
    assert(num_of_increases_window(test_inputs) == 5);
}

int main() {
    test(); 

    std::vector<int> inputs = read_file("inputs/input1.txt");
    std::cout << num_of_increases(inputs) << std::endl;
    std::cout << num_of_increases_window(inputs) << std::endl;

    return 0;
}
